src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+#include <Functions/IFunctionDateOrDateTime.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+template <typename Transform>
+class FunctionDateOrDateTimeToDateTimeOrDateTime64 : public IFunctionDateOrDateTime<Transform>, WithContext
+{
+public:
+    const bool enable_extended_results_for_datetime_functions = false;
+
+    static FunctionPtr create(ContextPtr context_)
+    {
+        return std::make_shared<FunctionDateOrDateTimeToDateTimeOrDateTime64>(context_);
+    }
+
+    explicit FunctionDateOrDateTimeToDateTimeOrDateTime64(ContextPtr context_)
+        : WithContext(context_)
+        , enable_extended_results_for_datetime_functions(context_->getSettingsRef().enable_extended_results_for_datetime_functions)
+    {
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        this->checkArguments(arguments, /*is_result_type_date_or_date32*/ false);
+
+        const IDataType * from_type = arguments[0].type.get();
+        WhichDataType which(from_type);
+
+        std::string time_zone = extractTimeZoneNameFromFunctionArguments(arguments, 1, 0);
+
+        /// If the time zone is specified but empty, throw an exception.
+        /// only validate the time_zone part if the number of arguments is 2.
+        if (arguments.size() == 2 && time_zone.empty())
+            throw Exception(
+                "Function " + this->getName() + " supports a 2nd argument (optional) that must be non-empty and be a valid time zone",
```

Comment:
Are you sure that 'must be non-empty' makes sense?
Is not it better " ... (optional). If specified must be a valid time zone" or just "... (optional) that must be a valid time zone" ?

Reply:
I would second `... (optional) that must be a valid time zone`.

Reply:
done